kytos/core/buffers/factory.py

```diff
@@ -0,0 +1,93 @@
+"""Utilities for composing KytosEventBuffers"""
+from functools import reduce
+
+import limits
+import limits.storage as lstorage
+import limits.strategies as lstrategies
```

Comment:
@Ktmi, I was looking into `limits` source code and docs, and there's exclusive [`async/aio` strategies, check it out](https://github.com/alisaifee/limits/blob/master/limits/aio/strategies.py), also the [storage has a corresponding `async` abstraction](https://github.com/alisaifee/limits/blob/master/limits/aio/storage/memory.py) that's `asyncio.Task` based instead of using `treading.Timer`, so let's make sure we're also using `asyncio` all the way down where possible. As a result, later on when calling `hit` it'd be a bool coroutine `await moving_window.hit(one_per_minute, "test_namespace", "foo")`. 

Locally, I was exploring a 20/second limit, but then I realized that it was not reaching 20/seconds for a given dpid check out the graph below, in red it's filtering for one dpid, and in blue it's all pkt out messages in a ring topology that I was using, I also included some logs to see how it was sleeping, and even though in most cases it was only sleeping after fetching roughly 20 items or so (or hitting them 20), but in some cases it was only getting 10 more or less, looks like it's sleeping more than it should, so I'm going to recommend for you to research and double check this part again, and check if we can potentially use [the remaing value from `WindowStats` when using the `get_window_stats`](https://github.com/alisaifee/limits/blob/master/limits/aio/strategies.py#L106), if it actually returns the remaining time when it might also simplifies the current subtraction that's being computed. 

```
2023-10-17 16:10:19,006 - INFO [kytos.core.buffers.buffers] [buffers.py:162:aget] (MainThread) buffer msg_out will sleep for: 0.9936294555664062 secs, q len: 905, ids: (20 per 1 second,
 ('127.0.0.1', 57754))
2023-10-17 16:10:20,017 - INFO [kytos.core.buffers.buffers] [buffers.py:162:aget] (MainThread) buffer msg_out will sleep for: 0.9823873043060303 secs, q len: 877, ids: (20 per 1 second,
 ('127.0.0.1', 57754))
2023-10-17 16:10:21,011 - INFO [kytos.core.buffers.buffers] [buffers.py:162:aget] (MainThread) buffer msg_out will sleep for: 0.9889931678771973 secs, q len: 868, ids: (20 per 1 second,
 ('127.0.0.1', 57754))
2023-10-17 16:10:22,018 - INFO [kytos.core.buffers.buffers] [buffers.py:162:aget] (MainThread) buffer msg_out will sleep for: 0.9818587303161621 secs, q len: 856, ids: (20 per 1 second,
 ('127.0.0.1', 57754))
2023-10-17 16:10:23,018 - INFO [kytos.core.buffers.buffers] [buffers.py:162:aget] (MainThread) buffer msg_out will sleep for: 0.9812984466552734 secs, q len: 836, ids: (20 per 1 second,
 ('127.0.0.1', 57754))
```

![20231017_161154](https://github.com/kytos-ng/kytos/assets/1010796/39626823-a391-4b9e-b90c-75ae201eeab8)

Reply:
Yeah, I've seen the async version, but the reason I didn't use it because, then I would have to have two separate rate limiters between the async and non async version. However, it seems that we don't use non-async `get`, so I can change this to the pure async version if we go under the assumption that `get` is deprecated and that we only use `aget`.

Reply:
Right. Let's go with pure async then and leave the sync part unsupported. All [queue consumers on core are async](https://github.com/kytos-ng/kytos/blob/master/kytos/core/controller.py#L367-L376), they're either using [msg_out_event_handler](https://github.com/kytos-ng/kytos/blob/master/kytos/core/controller.py#L578) or the generic [event_handler](https://github.com/kytos-ng/kytos/blob/master/kytos/core/controller.py#L554).

Reply:
Switched the code to async. Not sure what's causing the rate to be halved like that.

Reply:
@Ktmi, I'm suspecting it might be the sleep interval, in some iterations if it sleeps more than expected then it makes sense that the rate halved. Maybe [the remaing value from WindowStats when using the get_window_stats](https://github.com/alisaifee/limits/blob/master/limits/aio/strategies.py#L106) might be helpful if you could look into it and then later on confirm with a local stress test too. I generated this graph with wireshark > IO graph sending a bunch of FlowMods. If you need help let me know.

Reply:
I'm not seeing the same behaviour you are. For the following graph, I multiplied the packets sent by of_lldp so it would send 192 to a single switch every 3 seconds with a limit of 100/second, and here are the results.
![image](https://github.com/kytos-ng/kytos/assets/46541367/9b3f0dc8-0b8c-4c9b-8618-58bc44665ce9)

Same conditions as above, but I have `of_lldp` try to send 400 packets every 3 seconds.

![image](https://github.com/kytos-ng/kytos/assets/46541367/d1c956fe-eec5-42b5-a238-b4911c087afa)

In both above graphs, their are periods where packets aren't being sent at the same throughput, but the max throughput isn't being limited in the same way you are showing.

As for the remaining value from window_stats, that's the amount of room left in the window, I'll add a guard to try again if the value is not 0.